vrx_ign/src/SimpleBuoyancy.cc

```diff
@@ -0,0 +1,358 @@
+    auto ptr = const_cast<sdf::Element *>(_sdf.get());
+    for (sdf::ElementPtr buoyancyElem = ptr->GetElement("buoyancy");
+        buoyancyElem;
+        buoyancyElem = buoyancyElem->GetNextElement("buoyancy"))
+    {
+      BuoyancyObject buoyObj;
+      buoyObj.Load(_entity, buoyancyElem, _ecm);
+      buoyObj.height = this->dataPtr->fluidLevel;
+
+      // Add buoyancy object to list and display stats.
+      igndbg << buoyObj << std::endl;
+      this->dataPtr->buoyancyObjects.push_back(std::move(buoyObj));
+    }
+  }
+
+  // Get the gravity from the world.
+  auto worldEntity = gazebo::worldEntity(_ecm);
+  auto world = gazebo::World(worldEntity);
+  auto gravityOpt = world.Gravity(_ecm);
+  if (!gravityOpt)
+  {
+    ignerr << "Unable to get the gravity from the world" << std::endl;
+    return;
+  }
+  this->dataPtr->gravity = *gravityOpt;
+}
+
+//////////////////////////////////////////////////
+void SimpleBuoyancy::PreUpdate(const gazebo::UpdateInfo &_info,
+    gazebo::EntityComponentManager &_ecm)
+{
+  IGN_PROFILE("SimpleBuoyancy::PreUpdate");
+
+  // Elapsed time since the last update.
+  double dt;
+  double simTime;
+  if (this->dataPtr->wavefield.Active())
+  {
+    simTime = std::chrono::duration<double>(_info.simTime).count();
+    dt = simTime - this->dataPtr->lastSimTime;
+    this->dataPtr->lastSimTime = simTime;
+  }
+
+  for (auto &buoyancyObj : this->dataPtr->buoyancyObjects)
+  {
+    // Link pose in world coordinates.
+    auto worldPoseComp = buoyancyObj.link.WorldPose(_ecm);
+    if (!worldPoseComp)
+    {
+      ignerr << "No worldpose component found" << std::endl;
+      continue;
+    }
+    math::Pose3d linkFrame = *worldPoseComp;
+
+    // Apply the offset.
+    linkFrame = linkFrame * buoyancyObj.pose;
+
+    // std::cerr << simTime << " " << linkFrame.Pos() << std::endl;
+
+    if (this->dataPtr->wavefield.Active())
+    {
+      // Compute the wave displacement at the offset point,
+      // relative to the mean water level.
+      double waveHeight = this->dataPtr->wavefield.ComputeDepthSimply(
+        linkFrame.Pos(), simTime);
+      buoyancyObj.waterSpeed = (waveHeight - buoyancyObj.height) / dt;
+      buoyancyObj.height = waveHeight;
+    }
+
+    auto submergedVolume = buoyancyObj.shape->CalculateVolume(linkFrame,
+      buoyancyObj.height + this->dataPtr->fluidLevel);
+
+    // Calculate buoyancy and drag forces.
+    if (submergedVolume.volume > 1e-6)
+    {
+      // By Archimedes' principle,
+      math::Vector3d buoyancy = -this->dataPtr->fluidDensity *
+        submergedVolume.volume * this->dataPtr->gravity;
+
+      float partialMass = buoyancyObj.mass * submergedVolume.volume
+        / buoyancyObj.shape->volume;
+
+      auto linVel = buoyancyObj.link.WorldLinearVelocity(_ecm);
+      if (!linVel)
+      {
+        ignerr << "No linear velocity component found" << std::endl;
+        continue;
+      }
+
+      // Drag (based on Exact Buoyancy for Polyhedra by Eric Catto).
```

Comment:
Recommend removing this citation and putting full citation at top of file.

Reply:
I added the full citation to the header file in [19d536c](https://github.com/osrf/vrx/pull/500/commits/19d536c20c950b825ac3c5bb0ee970eebddbfdce).